File: src/ast.py
from __future__ import annotations
from enum import StrEnum
from typing import Literal
from dataclasses import dataclass, field
from .tokens import Token

class CDDLNode:
    '''
    Abstract base class for all nodes in the abstract syntax tree.
    '''
    def _str(self) -> str:
        '''
        Function must be implemented in all subclasses
        '''
        raise Exception('_str method must be implemented in subclass')

    def str(self) -> str:
        return self._str()

class WrappedNode(CDDLNode):
    '''
    A wrapped node is a node optionally enclosed in an open and close token.
    '''
    openToken: Token | None = None
    closeToken: Token | None = None

    def str(self) -> str:
        output = ''
        output += self.openToken.str() if self.openToken is not None else ''
        output += self._str()
        output += self.closeToken.str() if self.closeToken is not None else ''
        return output

class TokenNode(WrappedNode):
    '''
    A token node is a node that essentially represents a concrete token and/or
    that may be part of a list.

    It stores the comments and whitespaces that may come *before* it, and an
    optional separator token that may be used *after* it to separate it from
    the next token in an underlying list.

    The separator remains None when the node is not part of a list, or not part
    of a list that uses separators.

    A token node is a wrapped node if its openToken and closeToken properties
    are set.
    '''
    # Comments and whitespace *before* the node
    comments: list[Token] = []
    whitespace: str = ''
    separator: Token | None = None

    def __init__(self) -> None:
        self.comments = []
        self.whitespace = ''
        self.separator = None

    def _prestr(self) -> str:
        '''
        Function may be useful in subclasses to output something
        before the comments and whitespace associated with the
        main token
        '''
        return ''

    def str(self) -> str:
        output = self._prestr()
        for comment in self.comments:
            output += comment.str()
        output += self.whitespace
        output += super().str()
        if self.separator is not None:
            output += self.separator.str()
        return output

    def setComments(self, token: Token) -> None:
        self.comments = token.comments
        self.whitespace = token.whitespace

@dataclass
class CDDLTree(TokenNode):
    '''
    Represents a set of CDDL rules
    '''
    rules: list[Rule]

    def __post_init__(self):
        super().__init__()

    def _str(self) -> str:
        return ''.join([item.str() for item in self.rules])

@dataclass
class Rule(CDDLNode):
    '''
    A group definition
    ```
    person = {
        age: int,
        name: tstr,
        employer: tstr,
    }
    ```
    '''
    name: Typename
    # Note: Consider storing as more directly useful booleans instead of as an
    # ASSIGN, TCHOICEALT or GCHOICEALT token (needed for spaces and comments)
    assign: Token
    type: Type | GroupEntry

    def __post_init__(self):
        super().__init__()

    def _str(self) -> str:
        return self.name.str() + self.assign.str() + self.type.str()

@dataclass
class GroupEntry(TokenNode):
    '''
    A group entry
    '''
    occurrence: Occurrence | None
    key: Memberkey | None
    type: Type | Group

    def __post_init__(self):
        super().__init__()

    def _str(self):
        output = ''
        if self.occurrence is not None:
            output += self.occurrence.str()
        if self.key is not None:
            output += self.key.str()
        output += self.type.str()
        return output

@dataclass
class Group(TokenNode):
    '''
    A group, meaning a list of group choices wrapped in parentheses or curly
    braces
    '''
    groupChoices: list[GroupChoice]
    isMap: bool

    def __post_init__(self):
        super().__init__()

    def _str(self) -> str:
        return ''.join([item.str() for item in self.groupChoices])

@dataclass
class GroupChoice(TokenNode):
    '''
    A group choice
    '''
    groupEntries: list[GroupEntry]

    def __post_init__(self):
        super().__init__()

    def _str(self) -> str:
        return ''.join([item.str() for item in self.groupEntries])

@dataclass
class Array(TokenNode):
    '''
    An array
    ```
    [ city: tstr ]
    ```
    '''
    groupChoices: list[GroupChoice]

    def __post_init__(self):
        super().__init__()

    def _str(self) -> str:
        return ''.join([item.str() for item in self.groupChoices])


@dataclass
class Tag(TokenNode):
    '''
    A tag definition
    ```
    #6.32(tstr)
    ```
    '''
    # TODO: consider storing the numeric part as an int or float instead of as
    # a NUMBER or FLOAT token (using Token for spaces and comments)
    numericPart: Token | None = None
    # TODO: Consider getting back to a Type, storing spaces and comments before
    # closing ")" separately, because a Group is fairly verbose
    # (use of a group means one can only access the type through:
    # typePart.groupChoices[0].groupEntries[0].type)
    typePart: Group | None = None

    def __post_init__(self):
        super().__init__()

    def _str(self) -> str:
        output = '#'
        output += self.numericPart.str() if self.numericPart is not None else ''
        output += self.typePart.str() if self.typePart is not None else ''
        return output

@dataclass
class Occurrence(TokenNode):
    n: int | float
    m: int | float
    # TODO: ideally, we wouldn't have the parser store tokens on top of
    # the min and max, that's just a quick and dirty way to get the
    # different combinations right
    tokens: list[Token] = field(default_factory=list)

    def __post_init__(self):
        super().__init__()

    def _str(self) -> str:
        return ''.join([item.str() for item in self.tokens])

@dataclass
class Value(TokenNode):
    '''
    A value (number, text or bytes)
    '''
    value: str
    type: Literal['number', 'text', 'bytes', 'hex', 'base64']

    def __post_init__(self):
        super().__init__()

    def _str(self) -> str:
        if self.type == 'number':
            return self.value
        elif self.type == 'text':
            return '"' + self.value + '"'
        elif self.type == 'bytes':
            return '\'' + self.value + '\''
        elif self.type == 'hex':
            return 'h\'' + self.value + '\''
        else:
            return 'b64\'' + self.value + '\''

@dataclass
class Typename(TokenNode):
    '''
    A typename (or groupname)
    '''
    name: str
    unwrapped: Token | None
    parameters: GenericParameters | GenericArguments | None = None

    def __post_init__(self):
        super().__init__()

    def _prestr(self) -> str:
        return self.unwrapped.str() if self.unwrapped is not None else ''

    def _str(self) -> str:
        output = self.name
        if self.parameters is not None:
            output += self.parameters.str()
        return output

@dataclass
class Reference(TokenNode):
    '''
    A reference to another production
    '''
    target: Group | Typename

    def __post_init__(self):
        super().__init__()

    def _str(self) -> str:
        return '&' + self.target.str()

# A type2 production is one of a few possibilities
Type2 = Value | Typename | Group | Array | Reference | Tag

@dataclass
class Range(TokenNode):
    '''
    A Range is a specific kind of Type1.

    The grammar allows a range boundary to be a Type2. In practice, it can only
    be an integer, a float, or a reference to a value that holds an integer or
    a float.
    '''
    min: Value | Typename
    max: Value | Typename
    # TODO: would be better to store that as an "inclusive" bool. Using a Token
    # for now to store spaces and comments
    rangeop: Token

    def __post_init__(self):
        super().__init__()

    def _str(self) -> str:
        return self.min.str() + self.rangeop.str() + self.max.str()

'''
Known control operators

Note: We may want to relax the check, control operators provide an extension
point for specs that define CDDL and they may define their own operators.
'''
OperatorName = Literal[
    # Control operators defined in the main CDDL spec
    'and', 'bits', 'cbor', 'cborseq', 'default',
    'eq', 'ge', 'gt', 'le', 'lt', 'ne',
    'regexp', 'size', 'within',

    # Control operators defined in RFC9165:
    # https://datatracker.ietf.org/doc/html/rfc9165
    'plus', 'cat', 'det', 'abnf', 'abnfb', 'feature',

    # proposed in the freezer:
    # https://datatracker.ietf.org/doc/html/draft-bormann-cbor-cddl-freezer-14#name-control-operator-pcre
    'pcre'
]

@dataclass
class Operator(TokenNode):
    '''
    An operator is a specific type of Type1
    '''
    type: Type2
    # TODO: Consider storing operator as str. Using Token for spaces and
    # comments but Token is of type CTLOP and literal in OperatorName
    name: Token
    controller: Type2

    def __post_init__(self):
        super().__init__()

    def _str(self):
        return self.type.str() + self.name.str() + self.controller.str()

# A Type1 production is either a Type2, a Range or an Operator
Type1 = Type2 | Range | Operator


@dataclass
class Memberkey(CDDLNode):
    type: Type1
    hasCut: bool
    # TODO: ideally, we wouldn't have the parser store tokens on top of
    # the type and hasCut, that's just a quick and dirty way to get the
    # different combinations of cut tokens right (with spaces and comments)
    tokens: list[Token] = field(default_factory=list)

    def __post_init__(self):
        super().__init__()

    def _str(self) -> str:
        return self.type.str() + ''.join([item.str() for item in self.tokens])

@dataclass
class Type(CDDLNode):
    '''
    A Type is a list of Type1, each representing a possible choice.
    '''
    types: list[Type1]

    def __post_init__(self):
        super().__init__()

    def _str(self) -> str:
        return ''.join([item.str() for item in self.types])

@dataclass
class GenericParameters(WrappedNode):
    '''
    A set of generic parameters
    '''
    parameters: list[Typename]

    def __post_init__(self):
        super().__init__()

    def _str(self) -> str:
        return ''.join([item.str() for item in self.parameters])

@dataclass
class GenericArguments(WrappedNode):
    '''
    A set of generic arguments
    '''
    parameters: list[Type1]

    def __post_init__(self):
        super().__init__()

    def _str(self) -> str:
        return ''.join([item.str() for item in self.parameters])

File: typescript/src/ast.ts
import { Token, Tokens } from "./tokens.ts";
import type { Serializable } from "./tokens.ts";

// Possible value types
export type ValueType = "number" | "text" | "bytes" | "hex" | "base64";

// Known control operators
export type OperatorName =
  | "and"
  | "bits"
  | "cbor"
  | "cborseq"
  | "default"
  | "eq"
  | "ge"
  | "gt"
  | "le"
  | "lt"
  | "ne"
  | "regexp"
  | "size"
  | "within"
  | "plus"
  | "cat"
  | "det"
  | "abnf"
  | "abnfb"
  | "feature"
  | "pcre";

// Prelude types defined in RFC810
export type PreludeType =
  | "any"
  | "uint"
  | "nint"
  | "int"
  | "bstr"
  | "bytes"
  | "tstr"
  | "text"
  | "tdate"
  | "time"
  | "number"
  | "biguint"
  | "bignint"
  | "bigint"
  | "integer"
  | "unsigned"
  | "decfrac"
  | "bigfloat"
  | "eb64url"
  | "eb64legacy"
  | "eb16"
  | "encoded-cbor"
  | "uri"
  | "b64url"
  | "b64legacy"
  | "regexp"
  | "mime-message"
  | "cbor-any"
  | "float16"
  | "float32"
  | "float64"
  | "float16-32"
  | "float32-64"
  | "float"
  | "false"
  | "true"
  | "bool"
  | "nil"
  | "null"
  | "undefined";

export type Markup = [string | null, string | null];

export class Marker {
  /**
   * Serialize a Token.
   *
   * The function must handle whitespaces and comments that the Token
   * contains.
   */
  serializeToken(token: Token, node: CDDLNode): string {
    return token.serialize();
  }

  /**
   * Serialize a Value.
   */
  serializeValue(prefix: string, value: string, suffix: string, node: Value): string {
    return prefix + value + suffix;
  }

  /**
   * Serialize a typename or a groupname
   */
  serializeName(name: string, node: Typename): string {
    return name;
  }

  /**
   * Wrapping markup for a node as a whole if needed
   */
  markupFor(node: CDDLNode): Markup {
    return [null, null];
  }
}

/**
 * Abstract base class for all nodes in the abstract syntax tree.
 */
export abstract class CDDLNode {
  parentNode: CDDLNode | null = null;

  serialize(marker?: Marker): string {
    // Make sure that parentNode relationships are properly set
    this.setChildrenParent();
    if (marker) {
      const markup = marker.markupFor(this);
      let output = markup[0] !== null ? markup[0] : "";
      output += this._serialize(marker);
      output += markup[1] !== null ? markup[1] : "";
      return output;
    }
    return this._serialize();
  }

  setChildrenParent(): void {
    /**
     * Initialize the parentNode links from children nodes to this node
     * so that marker can access and adapt its behavior based on the
     * current context.
     */
    for (const child of this.getChildren()) {
      child.parentNode = this;
      child.setChildrenParent();
    }
  }

  getChildren(): CDDLNode[] {
    /**
     * Return the list of children nodes attached to this node
     */
    return [];
  }

  /**
   * Function must be implemented in all subclasses.
   */
  abstract _serialize(marker?: Marker): string;

  protected _serializeToken(token: Token | null | undefined, marker?: Marker): string {
    if (!token) {
      return "";
    }
    if (!marker) {
      return token.serialize();
    }
    return marker.serializeToken(token, this);
  }

  toString(indent: number = 0): string {
    return "  ".repeat(indent) + this.constructor.name;
  }

  toJSON(): Serializable {
    return {};
  }
}

/**
 * A wrapped node is a node optionally enclosed in an open and close token.
 */
export abstract class WrappedNode extends CDDLNode {
  openToken: Token | null = null;
  closeToken: Token | null = null;

  override serialize(marker?: Marker): string {
    let output = "";
    if (marker) {
      const markup = marker.markupFor(this);
      output += markup[0] !== null ? markup[0] : "";
    }
    output += this._serializeToken(this.openToken, marker);
    output += this._serialize(marker);
    output += this._serializeToken(this.closeToken, marker);
    if (marker) {
      const markup = marker.markupFor(this);
      output += markup[1] !== null ? markup[1] : "";
    }
    return output;
  }

  override toString(indent: number = 0): string {
    const res: string[] = [super.toString(indent)];
    if (this.openToken) {
      res.push(this.openToken.toString(indent + 1));
    }
    if (this.closeToken) {
      res.push(this.closeToken.toString(indent + 1));
    }
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      openToken: this.openToken?.toJSON(),
      closeToken: this.closeToken?.toJSON()
    });
  }
}

/**
 * A token node is a node that essentially represents a concrete token and/or
 * that may be part of a list.
 *
 * It stores the comments and whitespaces that may come *before* it, and an
 * optional separator token that may be used *after* it to separate it from
 * the next token in an underlying list.
 *
 * The separator remains null when the node is not part of a list, or not part
 * of a list that uses separators.
 *
 * A token node is a wrapped node if its openToken and closeToken properties
 * are set.
 */
export abstract class TokenNode extends WrappedNode {
  // Comments and whitespace *before* the node
  comments: Token[] = [];
  whitespace: string = "";
  separator: Token | null = null;

  /**
   * Function may be useful in subclasses to output something
   * before the comments and whitespace associated with the
   * main token
   */
  protected _prestr(marker?: Marker): string {
    return "";
  }

  override serialize(marker?: Marker): string {
    let output = this._prestr(marker);
    for (const comment of this.comments) {
      output += this._serializeToken(comment, marker);
    }
    output += this.whitespace;
    output += super.serialize(marker);
    output += this._serializeToken(this.separator, marker);
    return output;
  }

  setComments(token: Token): void {
    this.comments = token.comments;
    this.whitespace = token.whitespace;
  }

  override toString(indent: number = 0): string {
    const res: string[] = [super.toString(indent)];
    for (const comment of this.comments) {
      res.push(comment.toString(indent + 1));
    }
    if (this.whitespace !== "") {
      res.push("  ".repeat(indent + 1) + `whitespaces: ${this.whitespace.length}`);
    }
    if (this.separator) {
      res.push(this.separator.toString(indent + 1));
    }
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      comments: this.comments.map(c => c.toJSON()),
      whitespace: this.whitespace,
      separator: this.separator?.toJSON()
    });
  }
}

/**
 * Represents a set of CDDL rules
 */
export class CDDLTree extends TokenNode {
  constructor(public rules: Rule[]) {
    super();
  }

  override getChildren(): CDDLNode[] {
    return this.rules;
  }

  _serialize(marker?: Marker): string {
    return this.rules.map((item) => item.serialize(marker)).join("");
  }

  override toString(indent: number = 0): string {
    const res: string[] = [super.toString(indent)];
    for (const rule of this.rules) {
      res.push(rule.toString(indent + 1));
    }
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      rules: this.rules.map(r => r.toJSON())
    });
  }
}

/**
 * A group definition
 */
export class Rule extends CDDLNode {
  constructor(
    public name: Typename,
    public assign: Token,
    public type: Type | GroupEntry
  ) {
    super();
  }

  override getChildren(): CDDLNode[] {
    return [this.name, this.type];
  }

  _serialize(marker?: Marker): string {
    let output = this.name.serialize(marker);
    output += this._serializeToken(this.assign, marker);
    output += this.type.serialize(marker);
    return output;
  }

  override toString(indent: number = 0): string {
    return [
      super.toString(indent),
      this.name.toString(indent + 1),
      this.assign.toString(indent + 1),
      this.type.toString(indent + 1),
    ].join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      name: this.name.toJSON(),
      assign: this.assign.toJSON(),
      type: this.type.toJSON()
    });
  }
}

/**
 * A group entry
 */
export class GroupEntry extends TokenNode {
  constructor(
    public occurrence: Occurrence | null,
    public key: Memberkey | null,
    public type: Type
  ) {
    super();
  }

  override getChildren(): CDDLNode[] {
    const children: CDDLNode[] = [];
    if (this.occurrence) {
      children.push(this.occurrence);
    }
    if (this.key) {
      children.push(this.key);
    }
    children.push(this.type);
    return children;
  }

  _serialize(marker?: Marker): string {
    let output = "";
    if (this.occurrence) {
      output += this.occurrence.serialize(marker);
    }
    if (this.key) {
      output += this.key.serialize(marker);
    }
    output += this.type.serialize(marker);
    return output;
  }

  /**
   * Return true if GroupEntry can be converted to a proper Type.
   */
  isConvertibleToType(): boolean {
    return (
      this.occurrence === null &&
      this.key === null &&
      (!(this.type instanceof Group) ||
        this.type instanceof Array ||
        this.type instanceof Map)
    );
  }

  override toString(indent: number = 0): string {
    const res: string[] = [super.toString(indent)];
    if (this.occurrence) {
      res.push(this.occurrence.toString(indent + 1));
    }
    if (this.key) {
      res.push(this.key.toString(indent + 1));
    }
    res.push(this.type.toString(indent + 1));
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      occurrence: this.occurrence?.toJSON(),
      key: this.key?.toJSON(),
      type: this.type.toJSON()
    });
  }
}

/**
 * A group, meaning a list of group choices wrapped in parentheses
 */
export class Group extends TokenNode {
  constructor(public groupChoices: GroupChoice[]) {
    super();
  }

  override getChildren(): CDDLNode[] {
    return this.groupChoices;
  }

  _serialize(marker?: Marker): string {
    return this.groupChoices.map((item) => item.serialize(marker)).join("");
  }

  override toString(indent: number = 0): string {
    const res: string[] = [super.toString(indent)];
    for (const item of this.groupChoices) {
      res.push(item.toString(indent + 1));
    }
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      groupChoices: this.groupChoices.map(g => g.toJSON())
    });
  }
}

/**
 * A map, meaning a list of group choices wrapped in curly braces
 */
export class Map extends Group {}

/**
 * An array
 */
export class Array extends Group {}

/**
 * A group choice
 */
export class GroupChoice extends TokenNode {
  constructor(public groupEntries: GroupEntry[]) {
    super();
  }

  override getChildren(): CDDLNode[] {
    return this.groupEntries;
  }

  _serialize(marker?: Marker): string {
    return this.groupEntries.map((item) => item.serialize(marker)).join("");
  }

  override toString(indent: number = 0): string {
    const res: string[] = [super.toString(indent)];
    for (const item of this.groupEntries) {
      res.push(item.toString(indent + 1));
    }
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      groupEntries: this.groupEntries.map(g => g.toJSON())
    });
  }
}

/**
 * A tag definition
 */
export class Tag extends TokenNode {
  constructor(
    public numericPart: Token | null = null,
    public typePart: Type | null = null
  ) {
    super();
  }

  override getChildren(): CDDLNode[] {
    return this.typePart ? [this.typePart] : [];
  }

  _serialize(marker?: Marker): string {
    let output: string = this._serializeToken(new Token(Tokens.HASH, ""), marker);
    output += this._serializeToken(this.numericPart, marker);
    output += this.typePart ? this.typePart.serialize(marker) : "";
    return output;
  }

  override toString(indent: number = 0): string {
    const res: string[] = [super.toString(indent) + " (#)"];
    if (this.numericPart) {
      res.push(this.numericPart.toString(indent + 1));
    }
    if (this.typePart) {
      res.push(this.typePart.toString(indent + 1));
    }
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      numericPart: this.numericPart?.toJSON(),
      typePart: this.typePart?.toJSON()
    });
  }
}

/**
 * Occurrence
 */
export class Occurrence extends TokenNode {
  constructor(
    public n: number,
    public m: number,
    public tokens: Token[] = []
  ) {
    super();
  }

  _serialize(marker?: Marker): string {
    return this.tokens.map((item) => this._serializeToken(item, marker)).join("");
  }

  override toString(indent: number = 0): string {
    const res: string[] = [super.toString(indent)];
    for (const item of this.tokens) {
      res.push(item.toString(indent + 1));
    }
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      n: this.n,
      m: this.m,
      tokens: this.tokens.map(t => t.toJSON())
    });
  }
}

/**
 * A value (number, text or bytes)
 */
export class Value extends TokenNode {
  constructor(
    public value: string,
    public type: ValueType
  ) {
    super();
  }

  _serialize(marker?: Marker): string {
    let prefix: string = "";
    let suffix: string = "";
    if (this.type === "text") {
      prefix = '"';
      suffix = '"';
    } else if (this.type === "bytes") {
      prefix = "'";
      suffix = "'";
    } else if (this.type === "hex") {
      prefix = "h'";
      suffix = "'";
    } else if (this.type === "base64") {
      prefix = "b64'";
      suffix = "'";
    }
    if (!marker) {
      return prefix + this.value + suffix;
    }
    return marker.serializeValue(prefix, this.value, suffix, this);
  }

  override toString(indent: number = 0): string {
    return "  ".repeat(indent) + `${this.constructor.name} (${this.type}): ${this.value}`;
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      value: this.value,
      type: this.type
    });
  }
}

/**
 * A typename (or groupname)
 */
export class Typename extends TokenNode {
  constructor(
    public name: string,
    public unwrapped: Token | null,
    public parameters: GenericParameters | GenericArguments | null = null
  ) {
    super();
  }

  override getChildren(): CDDLNode[] {
    return this.parameters ? [this.parameters] : [];
  }

  protected override _prestr(marker?: Marker): string {
    return this._serializeToken(this.unwrapped, marker);
  }

  _serialize(marker?: Marker): string {
    let output = "";
    if (!marker) {
      output = this.name;
    } else {
      output += marker.serializeName(this.name, this);
    }
    if (this.parameters) {
      output += this.parameters.serialize(marker);
    }
    return output;
  }

  override toString(indent: number = 0): string {
    const res: string[] = [super.toString(indent), "  ".repeat(indent + 1) + this.name];
    if (this.unwrapped) {
      res.push(this.unwrapped.toString(indent + 1));
    }
    if (this.parameters) {
      res.push(this.parameters.toString(indent + 1));
    }
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      name: this.name,
      unwrapped: this.unwrapped?.toJSON(),
      parameters: this.parameters?.toJSON()
    });
  }
}

/**
 * A choice built from a group (or a groupname)
 */
export class ChoiceFrom extends TokenNode {
  constructor(public target: Group | Typename) {
    super();
  }

  override getChildren(): CDDLNode[] {
    return [this.target];
  }

  _serialize(marker?: Marker): string {
    let output: string = this._serializeToken(new Token(Tokens.AMPERSAND, ""), marker);
    output += this.target.serialize(marker);
    return output;
  }

  override toString(indent: number = 0): string {
    const res: string[] = [
      super.toString(indent) + " (&)",
      this.target.toString(indent + 1),
    ];
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      target: this.target.toJSON()
    });
  }
}

/**
 * A type2 production is one of a few possibilities
 */
export type Type2 = Value | Typename | Type | Group | Map | Array | ChoiceFrom | Tag;

/**
 * A Range is a specific kind of Type1.
 */
export class Range extends TokenNode {
  constructor(
    public min: Value | Typename,
    public max: Value | Typename,
    public rangeop: Token
  ) {
    super();
  }

  override getChildren(): CDDLNode[] {
    return [this.min, this.max];
  }

  _serialize(marker?: Marker): string {
    let output = this.min.serialize(marker);
    output += this._serializeToken(this.rangeop, marker);
    output += this.max.serialize(marker);
    return output;
  }

  override toString(indent: number = 0): string {
    const res: string[] = [
      super.toString(indent),
      this.min.toString(indent + 1),
      this.rangeop.toString(indent + 1),
      this.max.toString(indent + 1),
    ];
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      min: this.min.toJSON(),
      max: this.max.toJSON(),
      rangeop: this.rangeop.toJSON()
    });
  }
}

/**
 * An operator is a specific type of Type1
 */
export class Operator extends TokenNode {
  constructor(
    public type: Type2,
    public name: Token,
    public controller: Type2
  ) {
    super();
  }

  override getChildren(): CDDLNode[] {
    return [this.type, this.controller];
  }

  _serialize(marker?: Marker): string {
    let output = this.type.serialize(marker);
    output += this._serializeToken(this.name, marker);
    output += this.controller.serialize(marker);
    return output;
  }

  override toString(indent: number = 0): string {
    const res: string[] = [
      super.toString(indent),
      this.type.toString(indent + 1),
      this.name.toString(indent + 1),
      this.controller.toString(indent + 1),
    ];
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      type: this.type.toJSON(),
      name: this.name.toJSON(),
      controller: this.controller.toJSON()
    });
  }
}

/**
 * A Type1 production is either a Type2, a Range or an Operator
 */
export type Type1 = Type2 | Range | Operator;

/**
 * Memberkey
 */
export class Memberkey extends CDDLNode {
  constructor(
    public type: Type1,
    public hasCut: boolean,
    public hasColon: boolean,
    public tokens: Token[] = []
  ) {
    super();
  }

  override getChildren(): CDDLNode[] {
    return [this.type];
  }

  _serialize(marker?: Marker): string {
    let output = this.type.serialize(marker);
    output += this.tokens.map((token) => this._serializeToken(token, marker)).join("");
    return output;
  }

  override toString(indent: number = 0): string {
    const res: string[] = [super.toString(indent), this.type.toString(indent + 1)];
    for (const item of this.tokens) {
      res.push(item.toString(indent + 1));
    }
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      type: this.type.toJSON(),
      hasCut: this.hasCut,
      hasColon: this.hasColon,
      tokens: this.tokens.map(t => t.toJSON())
    });
  }
}

/**
 * A Type is a list of Type1, each representing a possible choice.
 */
export class Type extends TokenNode {
  constructor(public types: Type1[]) {
    super();
  }

  override getChildren(): CDDLNode[] {
    return this.types;
  }

  _serialize(marker?: Marker): string {
    return this.types.map((item) => item.serialize(marker)).join("");
  }

  override toString(indent: number = 0): string {
    const res: string[] = [super.toString(indent)];
    for (const item of this.types) {
      res.push(item.toString(indent + 1));
    }
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      types: this.types.map(t => t.toJSON())
    });
  }
}

/**
 * A set of generic parameters
 */
export class GenericParameters extends WrappedNode {
  constructor(public parameters: Typename[]) {
    super();
  }

  override getChildren(): CDDLNode[] {
    return this.parameters;
  }

  _serialize(marker?: Marker): string {
    return this.parameters.map((item) => item.serialize(marker)).join("");
  }

  override toString(indent: number = 0): string {
    const res: string[] = [super.toString(indent)];
    for (const item of this.parameters) {
      res.push(item.toString(indent + 1));
    }
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      parameters: this.parameters.map(p => p.toJSON())
    });
  }
}

/**
 * A set of generic arguments
 */
export class GenericArguments extends WrappedNode {
  constructor(public parameters: Type1[]) {
    super();
  }

  override getChildren(): CDDLNode[] {
    return this.parameters;
  }

  _serialize(marker?: Marker): string {
    return this.parameters.map((item) => item.serialize(marker)).join("");
  }

  override toString(indent: number = 0): string {
    const res: string[] = [super.toString(indent)];
    for (const item of this.parameters) {
      res.push(item.toString(indent + 1));
    }
    return res.join("\n");
  }

  override toJSON(): Serializable {
    return Object.assign(super.toJSON(), {
      parameters: this.parameters.map(p => p.toJSON())
    });
  }
}
